### PR TITLE
[SPARK-45904][SQL][CONNECT] Mode function should supports sort with order direction

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Currently, the mode aggregate function only support sort desc with the input column.
But the mainstream database (e.g. H2 and Postgres) supports sort desc or asc.

https://github.com/apache/spark/pull/43663 will support `MODE() WITHIN GROUP (ORDER BY col [ASC|DESC])` syntax. 

### Why are the changes needed?
Mode function should supports sort with order direction.

### Does this PR introduce _any_ user-facing change?
'No'.
New feature.

### How was this patch tested?
Test cases updated.

### Was this patch authored or co-authored using generative AI tooling?
'No'.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala [hunk 1]
```diff
@@ -54,52 +51,28 @@ case class Mode(
     child: Expression,
     mutableAggBufferOffset: Int = 0,
     inputAggBufferOffset: Int = 0,
-    deterministicExpr: Expression = Literal.FalseLiteral)
+    isSortAsc: Option[Expression] = None)
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala:54]: I think it's not worth let `isSortAsc` as the child expression of `Mode` due to it's none or foldable.

**Comment by peter-toth** [comment]
Actually this is literally the same implementation I came up with internally before I added deterministic mode in https://github.com/apache/spark/pull/42755. But then we reduced the sort argument (`isSortAsc` in this PR) to a simple deterministic flag to keep the change simple.

@beliefer, are you referring to the `isSortAsc` argument of the mode function or are you referring to the new `MODE() WITHIN GROUP (ORDER BY col [ASC|DESC])` syntax by "mainstream databases support"?

**Comment by beliefer** [comment]
> @beliefer, are you referring to the `isSortAsc` argument of the mode function or are you referring to the new `MODE() WITHIN GROUP (ORDER BY col [ASC|DESC])` syntax by "mainstream databases support"?

Because I want add the new `MODE() WITHIN GROUP (ORDER BY col [ASC|DESC])` syntax by "mainstream databases supported. Please see https://github.com/apache/spark/pull/43663.

I found that the current implementation of `mode` only supports sort by column `ASC` if we make `deterministicExpr` is true.
We should let `mode` also supports by column `DESC`

**Comment by beliefer** [comment]
ping @MaxGekk cc @cloud-fan

**Comment by beliefer** [comment]
The GA failure is unrelated.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala [hunk 2]
```diff
@@ -842,19 +842,20 @@ object functions {
...
+   * If there are multiple values with the greatest frequency only one value will be returned. The
+   * value will be chosen based on optional sort direction. Use ascending order to get smallest
+   * value or descending order to get largest value from multiple values with the same frequency.
+   * If this clause is not specified the exact chosen value is not determined.
    *
    * @group agg_funcs
    * @since 4.0.0
    */
-  def mode(e: Column, deterministic: Boolean): Column = Column.fn("mode", e, lit(deterministic))
+  def mode(e: Column, isSortAsc: Boolean): Column = Column.fn("mode", e, lit(isSortAsc))
```
Comment (cloud-fan) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:858]: If we do want to support specify ordering, we should support ordering columns. I don't agree with this partial API change. Let's finish the SQL side first, then we can think about how the Scala API should look like.

**Comment by beliefer** [comment]
I will close due to https://github.com/apache/spark/pull/44184 merged.

**Pull request closed without merging**